core/src/main/scala/org/apache/spark/sql/execution/CachedPlanHelperExec.scala

```diff
@@ -97,7 +98,19 @@ case class CachedPlanHelperExec(childPlan: CodegenSupport, @transient session: S
   override def doConsume(ctx: CodegenContext, input: Seq[ExprCode], row: ExprCode): String =
     parent.doConsume(ctx, input, row)
 
-  override protected def doExecute(): RDD[InternalRow] = childPlan.execute()
+  override protected def doExecute(): RDD[InternalRow] = {
+    // This does not handle nested aggregates or joins. Should we handle ?
+    val repeatSeq = childPlan.collect {
+      case p: NonRecursivePlans => !p.producedForPlanInstance
```

Comment:
If I won't check for produced for plan instance it would again execute the same plan. Note that  CachedPlanHelperExec child is the plan itself. It's basically to break the recursion.